### PR TITLE
linux_testing: enable Rust by default if available

### PR DESCRIPTION
**Pull request by Ma27**
The first assumption[1] we had was that the `aarch64-unknown-none` target was missing from rustc and that this was the cause for the regression.

However, it turns out that the relevant code from `rustc` wasn't used anyways because the Makefile does `--sysroot /dev/null`[2] which prevents rustc from using its own libcore. So luckily we don't have to patch the Rust bootstrap to get aarch64-linux back working[3].

In fact, the Rust part seems broken for both 6.10 and 6.11[4], but I decided to not bother since none of those are longterm versions.

So what's left here?

* Enabling Rust for aarch64-linux because it clearly works[5].

* Turning off NFS_LOCALIO for aarch6y4-linux because breaks the build like this:

      /nix/store/f3k0rdhcd2cx57phx755c2xixgifw5m5-binutils-2.42/bin/ld: Unexpected GOT/PLT entries detected!
      /nix/store/f3k0rdhcd2cx57phx755c2xixgifw5m5-binutils-2.42/bin/ld: Unexpected run-time procedure linkages detected!
      /nix/store/f3k0rdhcd2cx57phx755c2xixgifw5m5-binutils-2.42/bin/ld: fs/nfs/localio.o: in function `nfs_local_iocb_alloc':
      /build/source/build/../fs/nfs/localio.c:290:(.text+0x324): undefined reference to `nfs_to'
      [...]

[1] https://github.com/NixOS/nixpkgs/pull/315121#issuecomment-2135805876 [2] https://lore.kernel.org/all/20231031201752.1189213-1-mmaurer@google.com/ [3] Of course I only realized this _after_ I spent some time hacking a rustc
    patch together 🙃
[4] This broke with

        error[E0463]: can't find crate for `core`
          |
          = note: the `aarch64-unknown-none` target may not be installed
          = help: consider downloading the target with `rustup target add aarch64-unknown-none`
          = help: consider building the standard library from source with `cargo build -Zbuild-std`

[5] While the build is fine, the VM tests are still panicking, but
    that's also the case for `kernel-generic` because of a 9p
    regression:

        switch_root: can't execute '/nix/store/zv87gw0yxfsslq0mcc35a99k54da9a4z-nixos-system-machine-test/init': Exec format error
        [    1.734997] Kernel panic - not syncing: Attempted to kill init! exitcode=0x00000100
        [    1.736002] CPU: 0 UID: 0 PID: 1 Comm: switch_root Not tainted 6.12.0-rc1 #1-NixOS
        [...]

    Reported as https://lore.kernel.org/all/D4LHHUNLG79Y.12PI0X6BEHRHW@mbosch.me/T/#u

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by K900** [comment]
What happens on other targets now that we unconditionally enable Rust?

**Comment by winterqt** [comment]
Could we just enable it on aarch64 and x86_64? I'm not sure if it even supports other arches rn.

File: pkgs/os-specific/linux/kernel/common-config.nix [hunk 1]
```diff
@@ -33,10 +33,8 @@ let
   kernelSupportsRust = lib.versionAtLeast version "6.7";
 
   # Currently only enabling Rust by default on kernel 6.12+,
-  # which actually has features that use Rust that we want,
-  # and only on x86_64, because of a nixpkgs rustc issue:
-  # https://github.com/NixOS/nixpkgs/pull/315121#issuecomment-2135805876
-  defaultRust = lib.versionAtLeast version "6.12" && stdenv.hostPlatform.isx86_64;
+  # which actually has features that use Rust that we want.
+  defaultRust = lib.versionAtLeast version "6.12";
```
Comment (winterqt) [inline pkgs/os-specific/linux/kernel/common-config.nix:37]:
```suggestion
  defaultRust = lib.versionAtLeast version "6.12" && (stdenv.hostPlatform.isx86_64 || stdenv.hostPlatform.isAarch64);
```

**Comment by Ma27** [comment]
Yeah good point, was too focused on getting the aarch64 build done :)

File: pkgs/os-specific/linux/kernel/common-config.nix [hunk 2]
```diff
@@ -641,7 +639,10 @@ let
       NFS_V4_1              = yes;  # NFSv4.1 client support
       NFS_V4_2              = yes;
       NFS_V4_SECURITY_LABEL = yes;
-      NFS_LOCALIO           = whenAtLeast "6.12" yes;
+
+      # Fails with
+      # `fs/nfs/localio.o: relocation R_AARCH64_ADR_PREL_PG_HI21 against symbol `nfs_to' which may bind externally can not be used when making a shared object; recompile with -fPIC`
+      NFS_LOCALIO           = lib.mkIf (!stdenv.hostPlatform.isAarch64) (whenAtLeast "6.12" yes);
```
Comment (winterqt) [inline pkgs/os-specific/linux/kernel/common-config.nix:645]: It's not clear to me why this only happens `withRust`? Or maybe I'm just misunderstanding and it's broken on aarch64 either way.

Comment (alyssais) [inline pkgs/os-specific/linux/kernel/common-config.nix:645]: Has this been reported?

Comment (Ma27) [reply pkgs/os-specific/linux/kernel/common-config.nix:645]: Thanks for the reminder. Didn't report it myself, but will check if others did so already.

**Comment by Ma27** [comment]
I pushed a potential fix by making the support for CONFIG_RUST by dsefault dependant on whether rustc supports the host platform.

For testing purposes, I decided to build it for powerpc64le (built successfully) and riscv64 which failed to build with

```
error: unknown argument: '-mno-riscv-attribute'
error: unsupported argument 'medany' to option '-mcmodel=' for target 'unknown'
error: unsupported option '-fpatchable-function-entry=4' for target 'unknown'
error: unknown target triple 'unknown'
```

This is known and apparently riscv specific: https://lore.kernel.org/lkml/31885EDD-EF6D-4EF1-94CA-276BA7A340B7@kernel.org/T/
Current recommendation is to not mix up GCC and rustc apparently. If you want me to, I can fix this in the upcoming days, but I'm not sure if we should delay a fix for a aarch64-linux, a tier1 architecture for that.

Anyways, is this (i.e. `availableOn`) the direction we want to go? cc @alyssais

File: pkgs/os-specific/linux/kernel/common-config.nix (hunk 2 above)
Comment (Ma27) [reply pkgs/os-specific/linux/kernel/common-config.nix:646]: Huh, apparently there's no report for that specific issue. Will do.

Comment (Ma27) [reply pkgs/os-specific/linux/kernel/common-config.nix:646]: Done as https://lore.kernel.org/all/D4OUJRP8YWRM.ATQ7KASTYX5H@mbosch.me/T/#u.

File: pkgs/os-specific/linux/kernel/generic.nix [hunk 3]
```diff
@@ -108,6 +108,7 @@ let
 
   commonStructuredConfig = import ./common-config.nix {
     inherit lib stdenv version;
+    rustAvailable = lib.meta.availableOn stdenv.hostPlatform rustc;
```
Comment (alyssais) [inline pkgs/os-specific/linux/kernel/generic.nix:111]: I don't think this is quite right — even if rustc is not available for the platform we're building for, rustc on the build platform might still be able to build Linux for it. I added a mechanism recently for doing this check properly — see gstreamer for an example.

**Review by alyssais — CHANGES_REQUESTED** [review]

**Comment by alyssais** [comment]
> Current recommendation is to not mix up GCC and rustc apparently. If you want me to, I can fix this in the upcoming days, but I'm not sure if we should delay a fix for a aarch64-linux, a tier1 architecture for that.

I wouldn't worry about this one for now.

**Comment by alyssais** [comment]
Although I guess if we're fixing the conditional anyway, might as well add a `!(stdenv.hostPlatform.isRiscV && stdenv.cc.isGNU)` check…

File: pkgs/os-specific/linux/kernel/generic.nix (hunk 3 above)
Comment (Ma27) [reply pkgs/os-specific/linux/kernel/generic.nix:111]: Ehh apologies, you even mentioned gstreamer part before already! How embarrassing :see_no_evil: 
Will take a look.

Comment (alyssais) [reply pkgs/os-specific/linux/kernel/generic.nix]: We shouldn't — I was doing it to 64-bit MIPS recently…

File: pkgs/os-specific/linux/kernel/common-config.nix [hunk 4]
```diff
@@ -641,7 +640,10 @@ let
...
```
Comment (alyssais) [inline pkgs/os-specific/linux/kernel/common-config.nix:646]: This should probably be in a separate commit, since it's unrelated to Rust?  Otherwise LGTM.

Comment (Ma27) [reply pkgs/os-specific/linux/kernel/common-config.nix:646]: Done!

**Comment by alyssais** [comment]
@OfBorg build linux_testing
